File: zemiao/714.py
class Solution(object):
    def maxProfit(self, prices, fee):
        """
        :type prices: List[int]
        :type fee: int
        :rtype: int
        """
        hold, release = -327680, -327680
        for price in prices:
            newRelease = max(release, hold+price-fee)
            hold = max(hold, release-price if release >=0 else -price)
            release = newRelease
            
        return max(release, 0)
        

File: ztuo/714.java
class Solution {
    public int maxProfit(int[] prices, int fee) {
        if(prices.length == 0) {
            return 0;
        }
        int profit = 0;
        int[] buy = new int[prices.length];
        int[] sell = new int[prices.length];
        buy[0] = -prices[0];
        for (int i = 1; i < prices.length; i++) {
            buy[i] = Math.max(buy[i-1], sell[i-1] - prices[i]);
            sell[i] = Math.max(sell[i-1], buy[i-1] + prices[i] - fee);
        }
        return sell[prices.length - 1];
    }
}